### ISSUE TITLE
Simultaneous Hangup Not Being Informed

### ISSUE THREAD

**Issue by andrew-paes** [AsterNET/AsterNET#257]
Hi, 

I know AsterNet is not having much support, but I would like some help on how to handle this bug.

When I'm finishing 2 or more calls at the same time, somehow on AsterNet, one of the calls are not handled. I turned debug messages on and it is clear AsterNet is logging just one of the HangupRequest on Manager Reader. No matter how long I wait I'll never see the other call HangupRequest mesage.

As you can see below:

>   Debug:[ManagerReader-47] ManagerReader:Run - Event: Newchannel ####
  Debug:[ManagerReader-47] ManagerReader:Run - Privilege: call,all
  Debug:[ManagerReader-47] ManagerReader:Run - Channel: SIP/Elastix TRUE Outgoing-00000073
  Debug:[ManagerReader-47] ManagerReader:Run - ChannelState: 0
  Debug:[ManagerReader-47] ManagerReader:Run - ChannelStateDesc: Down
  Debug:[ManagerReader-47] ManagerReader:Run - CallerIDNum: 51999887766
  Debug:[ManagerReader-47] ManagerReader:Run - CallerIDName: 51999887766
  Debug:[ManagerReader-47] ManagerReader:Run - ConnectedLineNum: <unknown>
  Debug:[ManagerReader-47] ManagerReader:Run - ConnectedLineName: <unknown>
  Debug:[ManagerReader-47] ManagerReader:Run - Language: en
  Debug:[ManagerReader-47] ManagerReader:Run - AccountCode:
  Debug:[ManagerReader-47] ManagerReader:Run - Context: from-trunk-sip-Elastix TRUE Outgoing
  Debug:[ManagerReader-47] ManagerReader:Run - Exten: 4800
  Debug:[ManagerReader-47] ManagerReader:Run - Priority: 1
  Debug:[ManagerReader-47] ManagerReader:Run - Uniqueid: 1623706898.115
  Debug:[ManagerReader-47] ManagerReader:Run - Linkedid: 1623706898.115
  Debug:[ManagerReader-47] ManagerReader:Run - 
  

> .....
  Debug:[ManagerReader-47] ManagerReader:Run - Event: Newchannel ####
  Debug:[ManagerReader-47] ManagerReader:Run - Privilege: call,all
  Debug:[ManagerReader-47] ManagerReader:Run - Channel: SIP/Elastix TRUE Outgoing-00000075
  Debug:[ManagerReader-47] ManagerReader:Run - ChannelState: 0
  Debug:[ManagerReader-47] ManagerReader:Run - ChannelStateDesc: Down
  Debug:[ManagerReader-47] ManagerReader:Run - CallerIDNum: 51999887755
  Debug:[ManagerReader-47] ManagerReader:Run - CallerIDName: 51999887755
  Debug:[ManagerReader-47] ManagerReader:Run - ConnectedLineNum: <unknown>
  Debug:[ManagerReader-47] ManagerReader:Run - ConnectedLineName: <unknown>
  Debug:[ManagerReader-47] ManagerReader:Run - Language: en
  Debug:[ManagerReader-47] ManagerReader:Run - AccountCode:
  Debug:[ManagerReader-47] ManagerReader:Run - Context: from-trunk-sip-Elastix TRUE Outgoing
  Debug:[ManagerReader-47] ManagerReader:Run - Exten: 4800
  Debug:[ManagerReader-47] ManagerReader:Run - Priority: 1
  Debug:[ManagerReader-47] ManagerReader:Run - Uniqueid: 1623706910.117
  Debug:[ManagerReader-47] ManagerReader:Run - Linkedid: 1623706910.117
  Debug:[ManagerReader-47] ManagerReader:Run - 

  

> .....
  Debug:[ManagerReader-47] ManagerReader:Run - Event: HangupRequest ####
  Debug:[ManagerReader-47] ManagerReader:Run - Privilege: call,all
  Debug:[ManagerReader-47] ManagerReader:Run - Channel: SIP/Elastix TRUE Outgoing-00000073
  Debug:[ManagerReader-47] ManagerReader:Run - ChannelState: 6
  Debug:[ManagerReader-47] ManagerReader:Run - ChannelStateDesc: Up
  Debug:[ManagerReader-47] ManagerReader:Run - CallerIDNum: 51999887766
  Debug:[ManagerReader-47] ManagerReader:Run - CallerIDName: 51999887766
  Debug:[ManagerReader-47] ManagerReader:Run - ConnectedLineNum: 131
  Debug:[ManagerReader-47] ManagerReader:Run - ConnectedLineName: TARM 131
  Debug:[ManagerReader-47] ManagerReader:Run - Language: en
  Debug:[ManagerReader-47] ManagerReader:Run - AccountCode:
  Debug:[ManagerReader-47] ManagerReader:Run - Context: from-internal-xfer
  Debug:[ManagerReader-47] ManagerReader:Run - Exten: 444
  Debug:[ManagerReader-47] ManagerReader:Run - Priority: 39
  Debug:[ManagerReader-47] ManagerReader:Run - Uniqueid: 1623706898.115
  Debug:[ManagerReader-47] ManagerReader:Run - Linkedid: 1623706898.115
  Debug:[ManagerReader-47] ManagerReader:Run - 

If I wait 3 seconds or more to finish the second call, everything works fine. 

What could be the possible issue?

Full log is [Simultaneous Hangup.txt](https://github.com/AsterNET/AsterNET/files/6651608/Simultaneous.Hangup.txt)

**Comment by Deantwo** [comment]
Which version of AsterNET are you using?
Which version of Asterisk are you using?
Can you show the code you use to make the HangupRequest call?

**Comment by andrew-paes** [comment]
Hi Dean, 

Thanks for the help.

> Which version of AsterNET are you using?

Assembly Version 1.3.0.0
I think that was 1.3.0, forked by me on 4th July 2019

> Which version of Asterisk are you using?

Asterisk 13 using CHANSIP Channel

> Can you show the code you use to make the HangupRequest call?

I'm using the Logger class in WinForm project gave as sample. The code is:

``` C#
#region Debug

public void Debug(object o)
{
    string caller = debugInfo();
    if (isVisible(MessageLevel.Debug, caller.GetHashCode()))
        writeLine("  Debug:", string.Concat(caller, " - ", o.ToString()));
}

public void Debug(string msg)
{
    string caller = debugInfo();
    if (isVisible(MessageLevel.Debug, caller.GetHashCode()))
        writeLine("  Debug:", string.Concat(caller, " - ", msg));
}

public void Debug(string format, params object[] args)
{
    string caller = debugInfo();
    if (isVisible(MessageLevel.Debug, caller.GetHashCode()))
    {
        string s = string.Concat(caller, " - ", string.Format(format, args));

        writeLine("  Debug:", s);
    }
}

public void Debug(string msg, Exception ex)
{
    string caller = debugInfo();
    if (isVisible(MessageLevel.Debug, caller.GetHashCode()))
        writeLine("  Debug:", string.Concat(caller, " - ", string.Format("{0}\n{1}", msg, ex)));
}

#endregion
```

**Comment by Deantwo** [comment]
> I'm using the Logger class in WinForm project gave as sample.

I was asking for your code where you are doing the HangupRequest call, not the logging method.

**Comment by andrew-paes** [comment]
> > I'm using the Logger class in WinForm project gave as sample.
> 
> I was asking for your code where you are doing the HangupRequest call, not the logging method.

``` C#
internal void Init()
{
    ManagerCon = new ManagerConnection(Address, Port, User, Password);

    ManagerCon.FireAllEvents = false; // Debug: faz exibir todos eventos via Manager_UnhandledEvent (armazena no log)
    ManagerCon.Hangup += Manager_Hangup;
    ManagerCon.HangupRequest += Manager_HangupRequest;

    ManagerCon.KeepAlive = true;
    ManagerCon.PingInterval = 10000;

    Connect();
}


private void Manager_Hangup(object sender, HangupEvent e)
{
    Debug.WriteLine(e.UniqueId, e.Attributes["linkedid"], e.CallerIdNum, e.Channel, e.ChannelStateDesc);
}

private void Manager_HangupRequest(object sender, HangupRequestEvent e)
{
    Debug.WriteLine(e.UniqueId, e.Attributes["linkedid"], e.CallerIdNum, e.Channel, e.ChannelStateDesc);
}
```

**Comment by Deantwo** [comment]
That is you making connection, and then again more logging.

Where are you sending the Hangup action to the Asterisk?

**Comment by andrew-paes** [comment]
> That is you making connection, and then again more logging.
> 
> Where are you sending the Hangup action to the Asterisk?

Well, I'm sorry if I wasn't clear. But I'm not sending any Hangup action. I'm just hanging up two phone calls at the same time, and somehow I am not receiving both messages, just one of them. Looks like one of the messages are just not being fired.

If I wait 3 seconds to hangup the second call from the first one, I do receive both messages.

Now I would like some help to figure out why I'm not receveing these concurrent events. I dont think Asterisk are scaping one of them. Could be some thread locking the second message?

**Comment by Deantwo** [comment]
Ah, so this is an [XY problem](https://en.wikipedia.org/wiki/XY_problem). You aren't trying to send a Hangup action, you are trying to detect when a phone changes from busy to available.

The `HangupRequest` event is raised when a `Hangup` AMI action is received. So this might not be what you are looking for. As far as I know at least, the Asterisk documentation is not at all clear on how this work.
See: https://wiki.asterisk.org/wiki/display/AST/Asterisk+13+ManagerEvent_HangupRequest

I would suggest you using the `NewState` event. If all you want is to know when a phone changes state from busy to available.
See: https://wiki.asterisk.org/wiki/display/AST/Asterisk+13+ManagerEvent_Newstate

**Comment by andrew-paes** [comment]
> Ah, so this is an [XY problem](https://en.wikipedia.org/wiki/XY_problem). You aren't trying to send a Hangup action, you are trying to detect when a phone changes from busy to available.
> 

Well sir, I didn't say that any moment, I said: "_When I'm **finishing** 2 or more calls at the same time, somehow on AsterNet, one of the calls are not handled._"

Any moment I said I was sending an Action command, that was your interpretation. Anyhow I'm sorry if I lead you to a mistake.

> The `HangupRequest` event is raised when a `Hangup` action is sent. So this not what you are looking for. As far as I know at least, the Asterisk documentation is so limited.
> See: https://wiki.asterisk.org/wiki/display/AST/Asterisk+13+ManagerEvent_HangupRequest
> 

Ok, that is a good point but I could grab the HangupRequest event even if I'm not sending a Hangup request action. It is not working fine when I'm finishing 2 or more calls at the same time. Also as you can see on log, I never sent any Hangup action but even like that I could generate a HangupRequest event. How?

> I would suggest you using the `NewState` event.
> See: https://wiki.asterisk.org/wiki/display/AST/Asterisk+13+ManagerEvent_Newstate

Thanks, I'll give that a shot but I dont think this is going to work very well. 

NewState event is fired just for channel's event but in this case, when a call is finished the channel is "closed", this is not coming back to _Down_ state. So if I don't have the hangup event and I don't have the down state for the channel how what you were thinking? I coudn't follow you.

**Comment by Deantwo** [comment]
I don't know how the `Hangup` event works from Asterisk, so maybe you should start by asking on the [Asterisk community forum](https://community.asterisk.org/). Once you know if the [`Hangup` AMI event](https://wiki.asterisk.org/wiki/display/AST/Asterisk+13+ManagerEvent_Hangup) is even sent as you think it is, we can check if it is an AsterNET issue. I doubt AsterNET is just randomly ignoring some of the events.

The log in your opening post clearly only shows one `Hangup` event (and for some reason a `HangupRequest` event) being received, so I don't think this is an AsterNET issue.

You can also try to monitor the AMI messages yourself if you want, for example with [Fiddler](https://www.telerik.com/fiddler) if you know how to use it. Then you can see exactly what the Asterisk server is sending and compare it to what AsterNET is receiving.

**Comment by andrew-paes** [comment]
> I don't know how the `Hangup` event works from Asterisk, so maybe you should start by asking on the [Asterisk community forum](https://community.asterisk.org/). Once you know if the [`Hangup` AMI event](https://wiki.asterisk.org/wiki/display/AST/Asterisk+13+ManagerEvent_Hangup) is even sent as you think it is, we can check if it is an AsterNET issue. I doubt AsterNET is just randomly ignoring some of the events.
> 
> The log in your opening post clearly only shows one `Hangup` event (and for some reason a `HangupRequest` event) being received, so I don't think this is an AsterNET issue.
> 
> You can also try to monitor the AMI messages yourself if you want, for example with [Fiddler](https://www.telerik.com/fiddler) if you know how to use it. Then you can see exactly what the Asterisk server is sending and compare it to what AsterNET is receiving.

Hi Dean,

Yeah, I logged all the messages sent by Asterisk AMI via telnet, and Asterisk is sending both mesages. I do have full log if you want to see, but you can try it by your self just doing telnet and logging in via AMI like this: http://the-asterisk-book.com/1.6/asterisk-manager-api.html

![image](https://user-images.githubusercontent.com/3381749/122770792-2d235500-d27c-11eb-8bd8-1934d624849e.png)
[Simultaneous Hangup Telnet Log.txt](https://github.com/AsterNET/AsterNET/files/6686967/Simultaneous.Hangup.Telnet.Log.txt)

I added "events:on"

```
telnet 10.0.4.150 5038
ACTION:LOGIN
ACTIONID: 1
USERNAME:admin
SECRET:mystrongpass
EVENTS:ON
```

Later I key pressed ENTER, sou I could see the log:
![image](https://user-images.githubusercontent.com/3381749/122771837-2e08b680-d27d-11eb-9e9b-a833e066020c.png)



What do you think it is? Threads? Buffer?

**Comment by Deantwo** [comment]
> [Simultaneous Hangup Telnet Log.txt](https://github.com/AsterNET/AsterNET/files/6686967/Simultaneous.Hangup.Telnet.Log.txt)

Did you have AsterNET running at the same time? Would be interesting to see the two logs side by side.

**Comment by andrew-paes** [comment]
> > [Simultaneous Hangup Telnet Log.txt](https://github.com/AsterNET/AsterNET/files/6686967/Simultaneous.Hangup.Telnet.Log.txt)
> 
> Did you have AsterNET running at the same time? Would be interesting to see the two logs side by side.

Yes I had, and I can say I was worried because a lot of messages are sent by AMI and it is not managed by AsterNet, at least it didn't show important messages like "SoftHangupRequest".

I'm looking for some explanation. I was wondering if it could be some buffer limit but I really don't know where else to look.

**Comment by Deantwo** [comment]
> at least it didn't show important messages like "SoftHangupRequest".

`SoftHangupRequest` isn't implemented in AsterNET at all., so that is a bad example. The only thing we are interested in right now is the `Hangup` event.

> I'm looking for some explanation. I was wondering if it could be some buffer limit but I really don't know where else to look.

I haven't a clue what the issue is without seeing the two logs to compare. But you are free to investigate on your own.

I know the network part of AsterNET is weird, but I have never figured out how it could be better written.

**Comment by andrew-paes** [comment]
> > at least it didn't show important messages like "SoftHangupRequest".
> 
> `SoftHangupRequest` isn't implemented in AsterNET at all., so that is a bad example. The only thing we are interested in right now is the `Hangup` event.
> 
> > I'm looking for some explanation. I was wondering if it could be some buffer limit but I really don't know where else to look.
> 
> I haven't a clue what the issue is without seeing the two logs to compare. But you are free to investigate on your own.
> 
> I know the network part of AsterNET is weird, but I have never figured out how it could be better written.

Well I did some tests and I found something a bit weird.

I deleted the "logger.Debug(line)" ManagerReader.Run after lieQueue.Deque and I put "logger.Debug(line)" inside ManagerReader.mrReaderCallbback and all ilnes were written as same as TCP on Prompt Cmd.

When I came it back as it was, some lines on a few tests weren't written. So we have something between these 2 steps, maybe on Thread pool or while locking the Queue, I really don't know what could possible be.

Furthermore, sometimes the HangupRequestEvent is not fired.

I will need to look at the code very carefully to figure out how to improve it. Maybe some middleware to handle the queue.

**Comment by Deantwo** [comment]
> Furthermore, sometimes the HangupRequestEvent is not fired.

Yeah like I said. The Asterisk documentation is horribly lacking in this area, so finding out how `HangupRequestEvent` and `HangupEvent` actually work first might be a good idea.

Anyway, I am curious as to what you will find! ^^
